java/org/contikios/cooja/Simulation.java

```diff
@@ -150,68 +155,133 @@ public Simulation(Cooja cooja, long seed) {
     this.cooja = cooja;
     randomGenerator = new SafeRandom(this);
     randomSeed = seed;
-  }
-
-  @Override
-  public void run() {
-    assert isRunning : "Did not set isRunning before starting";
-    lastStartRealTime = System.currentTimeMillis();
-    lastStartSimulationTime = getSimulationTimeMillis();
-    speedLimitLastRealtime = lastStartRealTime;
-    speedLimitLastSimtime = lastStartSimulationTime;
+    this.simulationThread = new Thread(() -> {
+      boolean isSimulationRunning = false;
+      boolean isAlive = true;
+      do {
+        EventQueue.Pair nextEvent = null;
+        try {
+          while (isAlive) {
+            Runnable cmd;
+            do {
+              cmd = isSimulationRunning ? commandQueue.poll() : commandQueue.take();
+              if (cmd == null) {
+                // No command to execute
+              } else if (cmd == START) {
+                if (!isSimulationRunning) {
+                  isSimulationRunning = true;
+                  setRunning(true);
+                }
+              } else if (cmd == STOP) {
+                if (isSimulationRunning) {
+                  isSimulationRunning = false;
+                  setRunning(false);
+                }
+              } else if (cmd == QUIT) {
+                isAlive = false;
+                isShutdown = true;
+                if (isSimulationRunning) {
+                  isSimulationRunning = false;
+                  setRunning(false);
```

Comment:
I think all three calls to setRunning can be combined into:
```Java
isAlive = cmd != QUIT;
isShutdown = cmd == QUIT;
isSimulationRunning = cmd == START;
setRunnning(cmd == START);
```

Reply:
`cmd` may also be any Runnable from, for example,`invokeSimulationThread`.

Reply:
I don't know the Java 17 syntax, but I think the code could look like this:
```Java
switch (cmd) {
  Integer i -> { isAlive = i != QUIT;
                 isShutdown = i == QUIT;
                 isSimulationRunning = i == START;
                 setRunnning(i == START);
  }
  Runnable r -> r.run();
}
```

Reply:
Nice! I have a lot to learn about latest Java. I will update this. Thanks!

Reply:
It looks like pattern matching switch is preview only in Java 17 but the calls have been combined as suggested.